openquake/hazardlib/gsim/campbell_bozorgnia_2014.py

```diff
@@ -346,16 +346,33 @@ class CampbellBozorgnia2014(GMPE):
     #: and depth (km) to the 2.5 km/s shear wave velocity layer (z2pt5)
     REQUIRES_SITES_PARAMETERS = {'vs30', 'z2pt5'}
 
-    #: Required rupture parameters are magnitude, rake, dip, ztor, rupture
+    #: Required rupture parameters are magnitude, rake, dip, rupture
     #: width and hypocentral depth
     REQUIRES_RUPTURE_PARAMETERS = {
-        'mag', 'rake', 'dip', 'ztor', 'width', 'hypo_depth'}
+        'mag', 'rake', 'dip', 'width', 'hypo_depth'}
 
     #: Required distance measures are Rrup, Rjb and Rx
     REQUIRES_DISTANCES = {'rrup', 'rjb', 'rx'}
 
+    REQUIRES_COMPUTED_PARAMETERS = {'ztor'}
+
     SJ = 0  # 1 for Japan
 
+    def set_parameters(self, rup):
+        """
+        Estimate the rupture parameters if not provided.
+        """
+        frv = np.zeros_like(rup.rake)
+        frv[(rup.rake > 30.) & (rup.rake < 150.)] = 1.
+
+        if not hasattr(rup, "ztor"):
```

Comment:
This tests is not required if you remove the ztor parameter from the REQUIRES_RUPTURE_PARAMETERS. Is this your preferred option? I think the cleanest solution would be to have two models, one that computes ztor automatically and one where ztor is computed by OQ for each rupture (i.e. the current behaviour).

Reply:
Hey,

The `REQUIRES_RUPTURE_PARAMETERS` no longer holds a parameter called `ztor`.
As the paper suggested, if the `ztor` is not provided, the model should be able to estimate the `ztor` from the given value, `mag.` but if it's suggested, we do not have to estimate them.
(`mag` is a required parameter, so we do not have to worry about not having `mag`.)

Hence, the `ztor` is now in a set called `REQUIRES_COMPUTED_PARAMETERS`.

[This](https://github.com/gem/oq-engine/pull/7554) was the idea that @micheles suggested.

Reply:
The only difference with the `in_cshm` from the link I provided above is that `in_cshm` always gets computed whereas `ztor` within this model, may or may not depends on whether a user has provided a `ztor` or not.

Reply:
Thanks for this note. Let me explain my point. In a hazard analysis 'ztor' is never provided as an input to the OQ Engine. 'ztor' is a parameter that the engine computes automatically for every rupture together with all the other parameters in REQUIRES_RUPTURE_PARAMETERS. If you remove 'ztor' from the REQUIRES_RUPTURE_PARAMETERS list, the Context object will never contain 'ztor' and your proposed change will always be used. 

As I briefly mentioned in my previous note, the cleanest way to add this possibility (i.e. of using the 'ztor' suggested by CB14) is to create a separate model e.g. CampbellBozorgnia2014NoZTor. The use can then decide to use the original version with 'ztor' computed from the rupture or this new version with 'ztor' inferred.

Reply:
So we use our own wrapper that runs the method called `get_mean_stds()` from `contexts` module
```python
# ctx here is a RuptureContext
results = contexts.get_mean_stds(model, ctx, [im])
```

and if the `RuptureContext` object has an attribute `ztor`, this statement will not get fired because `ctx` has one
![test-1](https://user-images.githubusercontent.com/7849113/166219407-36ad67b1-7002-41a9-8e3b-7f900094e949.PNG)

but if the `RuptureContext` object does not have one, then this statement gets triggered to estimate the `ztor` with the given `mag`.
![test-2](https://user-images.githubusercontent.com/7849113/166219519-29677126-242d-437a-94dd-9b131f2ae74c.PNG)
![test-3](https://user-images.githubusercontent.com/7849113/166219526-e62d3cd2-ed0c-4bce-ace7-ca702ea2f264.PNG)


I can see where the cleanest way of doing this comes from, and I agree. However, by splitting those, now we need to change the class name every time we want to use/test with/without `ztor` value. 
Currently, this PR is only about `ztor`, but my future PRs will include estimating the following parameters:
- width
- hypocentral depth

This code may not be the cleanest way of doing it, but from the user's point of view, they only need to care about the inputs they want to put.

E.g., We want to compute something with this model with `Ztor`, which we call `CampbellBozorgnia2014`, create a `RuptureContext`, and include `ztor`.
Then we now want to compute it without passing `Ztor`. All we have to do is, comment out the attribute called `ztor` instead of creating another class called `CampbellBozorgnia2014NoZtoR` or renaming the current class.

Reply:
```I can see where the cleanest way of doing this comes from, and I agree. However, by splitting those, now we need to change the class name every time we want to use/test with/without ztor value.``` This is not true if you set the way in which the model should operate at the instantiation level.